File: java/kindergarten-garden/src/main/java/KindergartenGarden.java
import java.util.Arrays;
import java.util.List;

public class KindergartenGarden {
  static final String[] DEFAULT_STUDENTS = {
    "Alice", "Bob", "Charlie", "David", "Eve", "Fred", "Ginny", "Harriet", "Ileana", "Joseph",
    "Kincaid", "Larry"
  };

  Plant[][] garden;
  List<String> students;

  KindergartenGarden(String plants) {
    this(plants, DEFAULT_STUDENTS);
  }

  KindergartenGarden(String plants, String[] studentArray) {
    String[] rows = plants.split("\n");
    garden = new Plant[rows.length][];
    for (int i = 0; i < garden.length; ++i) {
      garden[i] = new Plant[rows[i].length()];
      for (int j = 0; j < garden[i].length; ++j) {
        garden[i][j] = Plant.getPlant(rows[i].charAt(j));
      }
    }

    students = Arrays.stream(studentArray).sorted().toList();
  }

  List<Plant> getPlantsOfStudent(String student) {
    int index = students.indexOf(student);

    return List.of(
        garden[0][index * 2],
        garden[0][index * 2 + 1],
        garden[1][index * 2],
        garden[1][index * 2 + 1]);
  }
}

File: python/kindergarten-garden/kindergarten_garden.py
class Garden:
    _DEFAULT_STUDENTS = [
        'Alice',
        'Bob',
        'Charlie',
        'David',
        'Eve',
        'Fred',
        'Ginny',
        'Harriet',
        'Ileana',
        'Joseph',
        'Kincaid',
        'Larry'
    ]
    _CODE_TO_NAME = {
        'G': 'Grass',
        'C': 'Clover',
        'R': 'Radishes',
        'V': 'Violets'
    }

    def __init__(self, diagram, students=None):
        students = sorted(students if students is not None
                          else self._DEFAULT_STUDENTS)

        front, back = diagram.splitlines()
        codes = zip(front[::2], front[1::2], back[::2], back[1::2])

        self._student_to_plants = {student: [self._CODE_TO_NAME[code] for code in codes]
                                   for student, codes in zip(students, codes)}

    def plants(self, student):
        return self._student_to_plants[student]
